File: code-examples/js/memphis_examples/src/memphis_examples.ts
import { memphis, Memphis } from 'memphis-dev';

(async function () {
    let memphisConnection: Memphis | null = null;

    try {
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });

        const producer = await memphisConnection.producer({
            stationName: 'test_station',
            producerName: 'test_producer'
        });

            await producer.produce({
                message: Buffer.from("Hello world"), // you can also send a JS object - {}
            });

        memphisConnection.close();
    } catch (ex) {
        console.log(ex);
        if (memphisConnection) memphisConnection.close();
    }
})(); 

// These are examples of the async functiontions available by the Memphis API

// Connecting Examples
async function connectPassword(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
                host: "localhost",
                username: "root", // (root/application type user)
                password: "memphis" 
                });
    }catch(exception){
        // Handle exception
    }
}

async function connectToken(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
                host: "localhost",
                username: "root", // (root/application type user)
                connectionToken: "token"
                });
    }catch(exception){
        // Handle exception
    }
}

async function connectTLS(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
                host: "localhost",
                username: "root", // (root/application type user)
                keyFile: "~/tls_file_path.key",
                certFile: "~/tls_cert_file_path.crt",
                caFile: "~/tls_ca_file_path.crt"
                });
    }catch(exception){
        // Handle exception
    }
}

// Station Examples

async function stationDefault(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation"
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationRetentionType(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            retentionType: memphis.retentionTypes.MESSAGES
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationMemoryStorage(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            storageType: memphis.storageTypes.MEMORY
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithReplicas(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            replicas: 3
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationIdempotency(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            idempotencyWindowMs: 180000
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithSchema(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            schemaName: "sensorLogs"
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithDeadLetter(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            sendPoisonMsgToDls: true,
            sendSchemaFailedMsgToDls: false
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithDeadLetterToStation(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            sendPoisonMsgToDls: true,
            sendSchemaFailedMsgToDls: false,
            // dlsStation: "badSensorLogsStation" Coming soon
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithTieredStorage(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            tieredStorageEnabled: true
        });
    }catch(exception){
        // Handle exception
    }
}

async function stationWithPartitions(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.station({
            name: "myStation",
            partitionsNumber: 3
        });
    }catch(exception){
        // Handle exception
    }
}

// Attach Schema Examples

// This function is deprecated. Use enforceSchema instead

// Enforce Schema Examples

async function enforceSchema(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphis.enforceSchema({
            name: "mySchema",
            stationName: "myStation"
        });
    }catch(exception){
        // Handle exception
    }
}

// Detach Schema Examples

async function detachSchema(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.detachSchema({
            stationName: "myStation",
        });
    }catch(exception){
        // Handle exception
    }
}

// Close Examples

async function close(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.close();
    }catch(exception){
        // Handle exception
    }
}

// Produce Examples

async function producerBasic(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.producer({
            stationName: "myStation",
            producerName: "myNewProducer"
        });
    }catch(exception){
        // Handle exception
    }
}

async function produceFromConnection(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.produce({
            stationName: "myStation",
            producerName: "tempProducer",
            message: {some: "message"}
        });
    }catch(exception){
        // Handle exception
    }
}

async function produceAsync(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.produce({
            stationName: "myStation",
            producerName: "tempProducer",
            message: {some: "message"},
            ackWaitSec: 30,
            asyncProduce: true
        });
    }catch(exception){
        // Handle exception
    }
}

async function produceWithIdempotency(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.produce({
            stationName: "myStation",
            producerName: "tempProducer",
            message: {some: "message"},
            msgId: "42"
        });
    }catch(exception){
        // Handle exception
    }
}

async function produceWithHeaders(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });

        let headers = memphis.headers()
        headers.add("trace_header_example", "track_me_123")

        await memphisConnection.produce({
            stationName: "myStation",
            producerName: "tempProducer",
            message: {some: "message"},
            headers: headers
        });
    }catch(exception){
        // Handle exception
    }
}

async function produceWithPartition(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.produce({
            stationName: "myStation",
            producerName: "tempProducer",
            message: {some: "message"},
            producerPartitionKey: "somePartition"
        });
    }catch(exception){
        // Handle exception
    }
}

// Consume Examples

async function consumerDefualt(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.consumer({
            stationName: "myStation",
            consumerName: "newConsumer"
        });
    }catch(exception){
        // Handle exception
    }
}

async function consumerGroup(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.consumer({
            stationName: "myStation",
            consumerName: "newConsumer",
            consumerGroup: "consumerGroup1"
        });
    }catch(exception){
        // Handle exception
    }
}

async function consumerPollInterval(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.consumer({
            stationName: "myStation",
            consumerName: "newConsumer",
            pullIntervalMs: 2000
        });
    }catch(exception){
        // Handle exception
    }
}

async function consumerBatched(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.consumer({
            stationName: "myStation",
            consumerName: "newConsumer",
            pullIntervalMs: 10000,
            batchSize: 100,
            batchMaxTimeToWaitMs: 15000
        });
    }catch(exception){
        // Handle exception
    }
}

async function consumerMaxMessages(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.consumer({
            stationName: "myStation",
            consumerName: "newConsumer",
            pullIntervalMs: 10000,
            batchSize: 100,
            batchMaxTimeToWaitMs: 15000,
            maxMsgDeliveries: 100
        });
    }catch(exception){
        // Handle exception
    }
}

// Fetch Messages Examples

async function fetchMessages(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.fetchMessages({
            stationName: "myStation",
            consumerName: "tempConsumer",
            batchSize: 5,
            startConsumeFromSequence: 4
        });
    }catch(exception){
        // Handle exception
    }
}

// Create Schema Examples

async function createSchema(){
    let memphisConnection;
    try{
        memphisConnection = await memphis.connect({
            host: 'localhost',
            username: 'root',
            password: 'memphis',
        });
        await memphisConnection.createSchema({
            schemaName: "newSchema",
            schemaType: "json",
            schemaFilePath: "~/schemas/my_schema_path.json"
        });
    }catch(exception){
        // Handle exception
    }
}

File: code-examples/go/memphis_examples/memphis_examples.go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/memphisdev/memphis.go"
)

// These are examples of the functions available by the Memphis API

func main() {
	// Put methods here to run them...
}

// Connecting Examples

func ConnectPassword() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(conn.ConnId)
}

func ConnectToken() {
	conn, err := memphis.Connect("localhost", "root", memphis.ConnectionToken("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(conn.ConnId)
}

func ConnectTLS() {
	conn, err := memphis.Connect("localhost", "root", memphis.Tls(
		"~/tls_file_path.key",
		"~/tls_cert_file_path.crt",
		"~/tls_cert_file_path.crt",
	))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(conn.ConnId)
}

// Station Examples

func StationDefault() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation("myStation")

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationRetentionType() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.RetentionTypeOpt(memphis.Messages),
		memphis.RetentionVal(10),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationMemoryStorage() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.StorageTypeOpt(memphis.Memory),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationWithReplicas() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.Replicas(3),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationIdempotency() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.IdempotencyWindow(3*time.Minute),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationWithSchema() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.SchemaName("sensorLogs"),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationPoisonDls() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.SchemaName("SensorLogs"),
		memphis.SendPoisonMsgToDls(true),
		memphis.SendSchemaFailedMsgToDls(false),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationWithPoisonDlsToStation() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.SchemaName("SensorLogs"),
		memphis.SendPoisonMsgToDls(true),
		memphis.SendSchemaFailedMsgToDls(false),
		// memphis.DlsStation("badSensorMessagesStation"), When this gets released it'll work...
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationWithTieredStorage() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.TieredStorageEnabled(true),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

func StationWithPartitions() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	station, err := conn.CreateStation(
		"myStation",
		memphis.PartitionsNumber(3),
	)

	fmt.Printf(station.Name) // To quiet the go compiler...
}

// Attach Schema Examples

// This function is deprecated. Use enforce_schema instead

// Enforce Schema Examples

func EnforceSchema() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	err = conn.EnforceSchema(
		"SchemaName",
		"MyStation",
	)

	// Handle enforce schema err
}

// Detach Schema Examples

func DeatchSchema() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	err = conn.DetachSchema(
		"SchemaName",
	)

	// Handle enforce schema err
}

// Close Examples

func Close() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	conn.Close()

	// You can defer this after handling the connection error, if desired
}

// Produce Examples

func ProduceBasic() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	producer, err := conn.CreateProducer(
		"StationToProduceFor",
		"MyNewProducer",
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(producer.Name) // To quiet the go compiler...
}

func ProduceWithHeaders() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	producer, err := conn.CreateProducer(
		"StationToProduceFor",
		"MyNewProducer",
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	hdrs := memphis.Headers{}
	hdrs.New()
	err = hdrs.Add("key", "value")

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	err = producer.Produce(
		[]byte("My Message :)"),
		memphis.MsgHeaders(hdrs),
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}
}

func ProduceWithPartition() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	producer, err := conn.CreateProducer(
		"StationToProduceFor",
		"MyNewProducer",
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	err = producer.Produce(
		[]byte("My Message :)"),
		memphis.ProducerPartitionKey("2ndPartition"),
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}
}

// Consume Examples

func ConsumerDefault() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer(
		"MyStation",
		"MyNewConsumer",
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(consumer.Name) // To quiet the go compiler...
}

func ConsumerWithGroup() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer(
		"MyStation",
		"MyNewConsumer",
		memphis.ConsumerGroup("ConsumerGroup1"),
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(consumer.Name) // To quiet the go compiler...
}

func ConsumerPullInterval() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer(
		"MyStation",
		"MyNewConsumer",
		memphis.PullInterval(2*time.Second),
	)

	fmt.Printf(consumer.Name) // To quiet the go compiler...
}

func ConsumerBatchMaxTimeToWait() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer(
		"MyStation",
		"MyNewConsumer",
		memphis.PullInterval(10*time.Second),
		memphis.BatchSize(100),
		memphis.BatchMaxWaitTime(15*time.Second),
	)

	fmt.Printf(consumer.Name) // To quiet the go compiler...
}

func ConsumerMaxMsgDeliveries() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer(
		"MyStation",
		"MyNewConsumer",
		memphis.PullInterval(10*time.Second),
		memphis.BatchSize(100),
		memphis.BatchMaxWaitTime(15*time.Second),
		memphis.MaxMsgDeliveries(100),
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(consumer.Name) // To quiet the go compiler...
}

// Fetch Messages Examples

func FetchMessages() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	msgs, err := conn.FetchMessages(
		"MyStation",
		"MyConsumer",
		memphis.FetchBatchSize(5),
		// memphis.FetchStartCStartConsumeFromSequence(3), Why this not existing?
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(string(msgs[0].Data())) // To quiet the go compiler...
}

// Create Schema Examples

func CreateSchema() {
	conn, err := memphis.Connect("localhost", "root", memphis.Password("memphis"))

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	conn.CreateSchema(
		"SchemaName",
		"SchemaType",
		"~/PathToSchema.json",
	)

	if err != nil {
		fmt.Printf("An error has occured :( : %v", err)
		os.Exit(1)
	}

	fmt.Printf(conn.ConnId) // To quiet the go compiler...
}
